Feature/usage limit multiple

gmorice opened the pull request:
Issue #2902495: Adds usage limit per client for a Promotion.

Commit:
adds usage limit per client for a Promotion

GoZOo commented on modules/promotion/src/Plugin/Field/FieldWidget/UsageLimitPerClientWidget.php:
```diff
@@ -0,0 +1,77 @@
+  public function massageFormValues(array $values, array $form, FormStateInterface $form_state) {
+    $new_values = [];
+    foreach ($values as $key => $value) {
+      if (empty($form_state->getValue($this->parent)[$key]['limit'])) {
```
plutôt que de faire un continue au empty, nourri plutôt ta variable si !empty()

gmorice replied:
Corrigé

GoZOo commented on modules/promotion/src/Entity/Promotion.php:
```diff
@@ -428,6 +443,15 @@ public function available(OrderInterface $order) {
     }
+    if ($usage_limit_per_client = $this->getUsageLimitPerClient()) {
+      /** @var \Drupal\commerce_promotion\PromotionUsageInterface $usage */
+      $usage = \Drupal::service('commerce_promotion.usage');
```
Do not call services as static but use dependency injection

GoZOo replied:
Forget about it, i miss we are in  ContentEntityBase

gmorice replied:
ok

GoZOo commented on modules/promotion/src/Plugin/Field/FieldWidget/UsageLimitPerClientWidget.php:
```diff
@@ -0,0 +1,76 @@
+   * {@inheritdoc}
+   */
+  public function formElement(FieldItemListInterface $items, $delta, array $element, array &$form, FormStateInterface $form_state) {
+    $value = isset($items[$delta]->value) ? $items[$delta]->value : NULL;
```
Explain why you get data from parent and what does radio_path is used for

gmorice replied:
Comments added. 
I tried to be as clear as possible but do not hesitate to modify if necessary.

GoZOo committed:
Create usage_limit_per_client field storage definition

Commit:
Sniff fixes

Commit:
Change call to UsageLimitPerClientWidget parent's to get its 'limit' value.

Commit:
Don't define a default value to usage_limit_per_client widget.

Commit:
Fix checking of Promotion usage limit to take into account the eventual usage_limit_per_client.
Set default value to usage_limit_per_client widget to NULL during Promotion creation.
